### ISSUE TITLE
Suggestions are often invalid for proc-macro produced code

### ISSUE THREAD

**Issue by m-ou-se** [rust-lang/rust#85932]
The tokens generated by proc_macro (e.g. from `quote!()`) use `Span::call_site()`, which makes some suggestions produce invalid output:

```
error[E0308]: mismatched types
 --> src/main.rs:3:1
  |
3 | #[hello]
  | ^^^^^^^^
  | |
  | expected `&mut i32`, found integer
  | help: consider mutably borrowing here: `&mut #[hello]`
```

It's a bit tricky, because this call_site span does point into the current crate (to `#[hello]`), so this situaiton isn't detected in the same way as external `macro_rules`.

**Comment by m-ou-se** [comment]
Looks like `rustc_middle::lint::in_external_macro` should be used in more places. That one detects this `call_site` span by checking if `def_site` is a dummy or imported.

**Comment by m-ou-se** [comment]
I'm working on a fix specifically for these typeck/borrowing suggestions. Not sure how many other suggestions have the same problem.

**Comment by estebank** [comment]
@m-ou-se this is a prevalent problem. We need to add metadata to spans to distinguish call sites explicitly in an easy way and _always_ ignore suggestions in them.
